tools/src/main/scala/scala/scalanative/codegen/LLVMDebugInformation.scala

```diff
@@ -0,0 +1,375 @@
+      extends Type
+  case class DISubroutineType(types: Incr[DITypes]) extends Type
+  case class DITypes(retTpe: Option[Incr[Type]], arguments: Seq[Incr[Type]])
+      extends LLVMDebugInformation
+}
+
+object DebugInformationSection {
+  class Token {
+    def render(implicit dwf: DebugInformationSection.Builder) =
+      s"!${dwf.gidx.id(this)}"
+  }
+  class Builder(val gidx: GenIdx) {
+    private val b = collection.mutable.Map
+      .empty[Global, Incr[LLVMDebugInformation]]
+
+    private val anon =
+      collection.mutable.ListBuffer.empty[Incr[LLVMDebugInformation]]
+
+    private val globals = collection.mutable.Map
+      .empty[LLVMDebugInformation, Incr[LLVMDebugInformation]]
+
+    private val keyCache =
+      collection.mutable.Map.empty[Any, Incr[LLVMDebugInformation]]
+
+    def register[T <: LLVMDebugInformation](in: Global, dwarf: T): Incr[T] = {
+      val newTok = gidx.gen
+      b.synchronized {
+        if (b.contains(in))
+          throw new Exception(s"[dwarf] attempt to overwrite `$in` key")
+
+        val computed = Incr(newTok, dwarf)
+        b.update(in, computed)
+
+        computed
+      }
+    }
+
+    def put(in: Global, dwarf: Incr[LLVMDebugInformation]) = {
+      b.synchronized {
+        b.update(in, dwarf)
+      }
+    }
+
+    def fileFromPosition(
+        pos: Position
+    ): Incr[LLVMDebugInformation.DIFile] = {
+
+      // It's done in this rather verbose way because on Java 8 any approach using `zip` fails with
+      // value getOrElse is not a member of Iterable[scala.scalanative.codegen.llvm.Incr[Nothing]]
+      // Error:  possible cause: maybe a semicolon is missing before `value getOrElse'?
+
+      for {
+        filename <- pos.filename
+        dir <- pos.dir
+      } yield cachedBy[java.net.URI, DIFile](
+        pos.source,
+        LLVMDebugInformation.DIFile(filename, dir)
+      )
+    }
+      .getOrElse(
+        cachedBy[java.net.URI, DIFile](
+          pos.source,
+          LLVMDebugInformation.DIFile("unknown", "unknown")
+        )
+      )
+
+    def fileLocation(pos: Position): Incr[LLVMDebugInformation.DILocation] =
+      cachedBy(
+        pos,
+        LLVMDebugInformation.DILocation(
+          pos.line,
+          pos.column,
+          fileFromPosition(pos)
+        )
+      )
+    def scopeLocation(
+        pos: Position,
+        scope: Incr[LLVMDebugInformation.Scope]
+    ): Incr[LLVMDebugInformation.DILocation] =
+      cachedBy(
+        pos -> scope.tok,
+        LLVMDebugInformation.DILocation(pos.line, pos.column, scope)
+      )
+
+    def cached[T <: LLVMDebugInformation](in: LLVMDebugInformation): Incr[T] =
+      globals.synchronized {
+        globals.getOrElseUpdate(in, Incr(gidx.gen, in)).asInstanceOf[Incr[T]]
+      }
+
+    def cachedBy[K, T <: LLVMDebugInformation](
+        k: K,
+        in: => LLVMDebugInformation
+    )(implicit
+        ct: ClassTag[T]
+    ): Incr[T] =
+      keyCache.synchronized {
+        keyCache
```

Comment:
Maybe we can use a concurrent hash map so we can skip locking? Eg. `scala.collection.concurrent.TrieMap`

Reply:
Done in [f55c49b](https://github.com/scala-native/scala-native/pull/2869/commits/f55c49ba668729a24e5c252226f80f880e2f77f4)